reactor-core/src/main/java/reactor/core/publisher/Mono.java

```diff
@@ -3416,23 +3417,16 @@ public T block(Duration timeout) {
 	 * as long as the companion {@link Publisher} produces an onNext signal and the maximum number of repeats isn't exceeded.
 	 */
 	public final Mono<T> repeatWhenEmpty(int maxRepeat, Function<Flux<Long>, ? extends Publisher<?>> repeatFactory) {
-		return Mono.defer(() -> {
-			Flux<Long> iterations;
-
-			if(maxRepeat == Integer.MAX_VALUE) {
-				iterations = Flux.fromStream(LongStream.range(0, Long.MAX_VALUE).boxed());
+		return Mono.defer(() -> this.repeatWhen(o -> {
+			if (maxRepeat == Integer.MAX_VALUE) {
+				return repeatFactory.apply(o.index().map(Tuple2::getT1));
 			}
 			else {
-				iterations = Flux
-					.range(0, maxRepeat)
-					.map(Integer::longValue)
-					.concatWith(Flux.error(new IllegalStateException("Exceeded maximum number of repeats"), true));
+				return repeatFactory.apply(o.index().map(Tuple2::getT1)
+						.take(maxRepeat)
+						.concatWith(Flux.error(new IllegalStateException("Exceeded maximum number of repeats"), true)));
```

Comment:
@ericbottard  @simonbasle   
isn't it better to use `Flux.error(Supplier)` version?  
as it always creates Exception which captures stacktrace ?

Reply:
this could be an improvement to the existing code yeah, nothing to do with the bug addressed by this PR though. do you want to submit a PR @robotmrv ?

Reply:
@simonbasle   
created PR https://github.com/reactor/reactor-core/pull/2221